notifications/sinks.go

```diff
@@ -14,6 +14,51 @@ import (
 // externally, but others are suitable for export if the need arises. Albeit,
 // the tight integration with endpoint metrics should be removed.
 
+// SyncBroadcaster will synchronously send events to multiple sinks in serial order
+// If one of the sinks fail the overall operation fails
+type SyncBroadcaster struct {
+	sinks  []Sink
+	closed bool
+}
+
+// NewSyncBroadcaster creates a sink that will write to all the given sinks synchronously in serial order
+func NewSyncBroadcaster(sinks ...Sink) *SyncBroadcaster {
+	return &SyncBroadcaster{
+		sinks: sinks,
+	}
+}
+
+// Write writes the given events synchronously in serial order
+func (b *SyncBroadcaster) Write(events ...Event) error {
+	for _, sink := range b.sinks {
+		if err := sink.Write(events...); err != nil {
+			logrus.WithField("events", events).Errorf(
+				"syncbroadcaster: error writing events to %v, these events will be lost: %v", sink, err)
+			return err
```

Comment:
What does returning an error here mean for a request? It seems this is aiming for consistency, but is it expected that a client would need to retry if 1 of n sinks failed to send?

Reply:
Returning error here results in failure of HTTP request (500). The intention is client will retry the request.

Reply:
Retry but the original operation may have succeeded? Is there any case where the second request may go through a different code path because of the prior partial success?

Reply:
> Retry but the original operation may have succeeded?

Yes.

> Is there any case where the second request may go through a different code path because of the prior partial success?

Yes. In fact it happens currently where a blob push fails due to sink failure but the blob is actually stored. The client subsequently skips uploading since a `HEAD` check succeeds. This is an unfortunate flaw in the design.

Reply:
(e.g. engine pushes up a blob, blob gets written to storage backend but notifications fail, on second try engine does a HEAD request and sees blob is there?)